src/common/async/yield_context.h

```diff
@@ -21,25 +21,34 @@
 
 #include "acconfig.h"
 
+#include <optional>
+
 /// optional-like wrapper for a boost::asio::yield_context. operations that take
 /// an optional_yield argument will, when passed a non-empty yield context,
 /// suspend this coroutine instead of the blocking the thread of execution
-class optional_yield {
-  boost::asio::yield_context *y = nullptr;
- public:
-  /// construct with a valid io and yield_context
-  optional_yield(boost::asio::yield_context& y) noexcept : y(&y) {}
 
-  /// type tag to construct an empty object
-  struct empty_t {};
-  optional_yield(empty_t) noexcept {}
+namespace detail { using opt_t = std::optional<boost::asio::yield_context>; }
 
-  /// implicit conversion to bool, returns true if non-empty
-  operator bool() const noexcept { return y; }
+struct optional_yield final : detail::opt_t
+{
+ using detail::opt_t::opt_t;
```

Comment:
this allows construction from `std::nullopt`, so doesn't enforce use of the `null_yield` token

Reply:
*IF* we like this whole thing at all (which is, I think the real question), it will only ever hold a yield_context, and we've now defined null_yield to be std::nullopt, so making them having the same meaning doesn't overly bother me. But, we can also remove that ctor:
 optional_yield::optional_yield(std::nullopt_t) = delete;